crates/ra_hir_ty/src/_match.rs

```diff
@@ -0,0 +1,1174 @@
+    let res = match pat.as_pat(cx) {
+        Pat::Wild => None,
+        Pat::Tuple(pats) => Some(Constructor::Tuple { arity: pats.len() }),
+        Pat::Lit(lit_expr) => match cx.body.exprs[lit_expr] {
+            Expr::Literal(Literal::Bool(val)) => Some(Constructor::Bool(val)),
+            _ => return Err(MatchCheckNotImplemented),
+        },
+        Pat::TupleStruct { .. } | Pat::Path(_) => {
+            let pat_id = pat.as_id().expect("we already know this pattern is not a wild");
+            let variant_id =
+                cx.infer.variant_resolution_for_pat(pat_id).ok_or(MatchCheckNotImplemented)?;
+            match variant_id {
+                VariantId::EnumVariantId(enum_variant_id) => {
+                    Some(Constructor::Enum(enum_variant_id))
+                }
+                _ => return Err(MatchCheckNotImplemented),
+            }
+        }
+        _ => return Err(MatchCheckNotImplemented),
+    };
+
+    Ok(res)
+}
+
+fn all_constructors_covered(
+    cx: &MatchCheckCtx,
+    constructor: &Constructor,
+    used_constructors: &[Constructor],
+) -> bool {
+    match constructor {
+        Constructor::Tuple { arity } => {
+            used_constructors.iter().any(|constructor| match constructor {
+                Constructor::Tuple { arity: used_arity } => arity == used_arity,
+                _ => false,
+            })
+        }
+        Constructor::Bool(_) => {
+            if used_constructors.is_empty() {
+                return false;
+            }
+
+            let covers_true =
+                used_constructors.iter().any(|c| matches!(c, Constructor::Bool(true)));
+            let covers_false =
+                used_constructors.iter().any(|c| matches!(c, Constructor::Bool(false)));
+
+            covers_true && covers_false
+        }
+        Constructor::Enum(e) => cx.db.enum_data(e.parent).variants.iter().all(|(id, _)| {
+            for constructor in used_constructors {
+                if let Constructor::Enum(e) = constructor {
+                    if id == e.local_id {
+                        return true;
+                    }
+                }
+            }
+
+            false
+        }),
+    }
+}
+
+fn enum_variant_matches(cx: &MatchCheckCtx, pat_id: PatId, enum_variant_id: EnumVariantId) -> bool {
+    if let Some(VariantId::EnumVariantId(pat_variant_id)) =
+        cx.infer.variant_resolution_for_pat(pat_id)
+    {
+        if pat_variant_id.local_id == enum_variant_id.local_id {
+            return true;
+        }
+    }
+    false
+}
+
+#[cfg(test)]
+mod tests {
+    pub(super) use insta::assert_snapshot;
+    pub(super) use ra_db::fixture::WithFixture;
+
+    pub(super) use crate::test_db::TestDB;
+
+    pub(super) fn check_diagnostic_message(content: &str) -> String {
+        TestDB::with_single_file(content).0.diagnostics().0
+    }
+
+    pub(super) fn check_diagnostic_with_no_fix(content: &str) {
```

Comment:
the `_with_no_fix` is a bit superfluous since `hir` doesn't have a concept of fixes.

Reply:
Renamed to `check_diagnostic`. Let me know if you have ideas for a clearer name.